### PR TITLE
Test boto3 performance throughput

### PR DESCRIPTION
**Pull request by crusaderky**
Meter boto3 performance when reading a single large file from s3.
This is to have a baseline to compare potential fsspec problems against.

- XREF https://github.com/coiled/benchmarks/issues/821
- XREF https://github.com/coiled/general/issues/154
- XREF https://github.com/fsspec/s3fs/issues/677

### CODE REVIEW

**Comment by crusaderky** [comment]
CC @ntabris

File: tests/benchmarks/test_parquet.py [hunk 1]
```diff
@@ -1,17 +1,20 @@
 """
 Parquet-related benchmarks.
 """
+import io
 import uuid
 
+import boto3
```
Comment (crusaderky) [inline tests/benchmarks/test_parquet.py:7]: dependency of coiled

File: tests/benchmarks/test_parquet.py [hunk 2]
```diff
@@ -1,17 +1,20 @@
...
 import dask.dataframe as dd
 import dask.datasets
 import distributed
+import fsspec
```
Comment (crusaderky) [inline tests/benchmarks/test_parquet.py:11]: dependency of dask

**Comment by mrocklin** [comment]
ooh, I'm excited to see the results of this

**Comment by crusaderky** [comment]
I ran it 10 times and the results seem to conclude that

- there is no noticeable difference between boto3 and fsspec
- there is a "warm-up" period where s3 will become progressively faster, e.g. your data shifts towards the top of some frequently used cache - until the cache doesn't work for whatever reason and it slows down again
- some tests will be noisier than others just in virtue of where they are in the test suite - first test is always the noisiest because of the above point
- there is no difference in overall performance between reading the file in 5 chunks from 5 different workers vs just sequentially from a single worker
- decoding the dataframe with pyarrow takes a huge amount of time, and parallelizing the effort across 5 workers (`[dask]`) makes a difference

image

**Comment by crusaderky** [comment]
Ready for review and merge

**Comment by phofl** [comment]
> decoding the dataframe with pyarrow takes a huge amount of time, and parallelizing the effort across 5 workers ([dask]) makes a difference

I guess you are referring to the conversion from the Arrow Table to NumPy arrays? This is one of the best use-cases for the general purpose ``dtype_backend="pyarrow"`` in pandas 2.0. This is a no-copy op in this scenario and is pretty fast (I saw speedups up to 10 times the NumPy performance for the whole read_parquet operation in pandas).

Do you think it makes sense to add one benchmark for this case as well? Would have to use pandas 2.0 though

**Review by milesgranger — APPROVED** [review]
LGTM

**Comment by ntabris** [comment]
What do you think about having each test download the file N times (say, 4 or 8)? That might make tests more consistent since (I'm guessing) majority of those downloads will be after s3 has moved file to hot storage.

FWIW the question I was interested in is whether we'd get much better network throughput by using multiple concurrent downloads. I don't think this answers that.

**Comment by crusaderky** [comment]
> What do you think about having each test download the file N times (say, 4 or 8)? That might make tests more consistent since (I'm guessing) majority of those downloads will be after s3 has moved file to hot storage.

I thought about that, but I think we're already spending enough time on this test...

**Comment by ntabris** [comment]
> I thought about that, but I think we're already spending enough time on this test...

As in developer time or computer time?

**Comment by crusaderky** [comment]
> > I thought about that, but I think we're already spending enough time on this test...
> 
> As in developer time or computer time?

both

**Pull request merged**